spec.html

```diff
@@ -7875,6 +7880,251 @@ <h1>Forward Progress</h1>
+      </emu-note>
+      <emu-note>
+        WeakRef-obliviousness is defined on sets of objects instead of
+        individual objects to account for cycles. If it were defined on
+        individual objects, then an object in a cycle will be considered live
+        even though its Object value is only observed via WeakRefs of other
+        objects in the cycle.
+      </emu-note>
+
+      <p>
+        At any point during evaluation, a set of objects _S_ is
+        considered <dfn>live</dfn> if either of the following conditions is
+        met:
+      </p>
+
+      <ul>
+        <li>
+          Any element in _S_ is included in any agent's [[KeptAlive]] List.
+        </li>
+        <li>
+          There exists a valid future hypothetical WeakRef-oblivious execution
+          with respect to _S_ that observes the Object value of any object in _S_.
+        </li>
+      </ul>
+      <emu-note>
+        The intuition the second condition above intends to capture is that an
+        object is live if its identity is observable via non-WeakRef means. An
+        object's identity may be observed by observing a strict equality
+        comparison between objects or observing the object being used as key in
+        a Map.
+      </emu-note>
+      <emu-note>
+        <p>
+        Presence of an object in a field, an internal slot, or a property does
+        not imply that the object is live. For example if the object in
+        question is never passed back to the program, then it cannot be
+        observed.
+        </p>
+
+        <p>
+        This is the case for keys in a WeakMap, members of a WeakSet, as well
+        as the [[WeakRefTarget]] and [[UnregisterToken]] fields of a FinalizationRegistry
+        Cell record.
+        </p>
+
+        <p>
+        The above definition implies that, if a key in a WeakMap is not live,
+        then its corresponding value is not necessarily live either.
+        </p>
+      </emu-note>
+      <emu-note>
+        Liveness is the lower bound for guaranteeing which WeakRefs that
+        engines must not empty. In practice, liveness as defined here is
+        undecidable and engines use conservative approximations. There is
+        expected to be significant implementation leeway.
+      </emu-note>
+    </emu-clause>
+
+    <emu-clause id="sec-weakref-execution">
+      <h1>Execution</h1>
+
+      <p>At any time, if a set of objects _S_ is not live, an ECMAScript implementation may perform the following steps atomically:</p>
+
+      <emu-alg>
+        1. For each _obj_ of _S_, do
+          1. For each WeakRef _ref_ such that _ref_.[[WeakRefTarget]] is _obj_, do
+            1. Set _ref_.[[WeakRefTarget]] to ~empty~.
+          1. For each FinalizationRegistry _fg_ such that _fg_.[[Cells]] contains _cell_, such that _cell_.[[WeakRefTarget]] is _obj_, do
+            1. Set _cell_.[[WeakRefTarget]] to ~empty~.
+            1. Optionally, perform ! HostCleanupFinalizationRegistry(_fg_).
```

Comment:
The implementation of `HostCleanupFinalizationRegistry` is already free to not clean things up so I don't think this needs to be optional. 

```suggestion
            1. Perform ! HostCleanupFinalizationRegistry(_fg_).
```

Reply:
I disagree with this suggestion. While it is true that the host hook reserves the right to not do anything, calling out the optionality means hosts don't need to remember to explicitly allow optionality.

Reply:
Why don't we just explicitly call out the optionality in the hook then? Isn't the entire point of host hooks to abstract this kind of thing?

Reply:
It's important to call out optionality here to avoid a misunderstanding that it would be mandatory to trigger FinalizationRegistry. The word "optionally" makes the intended interpretation unambiguous, and it does not sacrifice precision/correctness. Remember that specifications are communication devices, not just mathematical artifacts. Because @devsnek's suggestion is an editorial one, we can continue considering it after this PR lands and possibly make the change later.

Reply:
As an editor I also prefer the "Optionally" phrasing.